bids/layout/core.py

```diff
@@ -0,0 +1,553 @@
+    def __iter__(self):
+        for i in self.unique():
+            yield(i)
+
+    def __deepcopy__(self, memo):
+        cls = self.__class__
+        result = cls.__new__(cls)
+        memo[id(self)] = result
+
+        for k, v in self.__dict__.items():
+            new_val = getattr(self, k) if k == 'regex' else deepcopy(v, memo)
+            setattr(result, k, new_val)
+        return result
+
+    def match_file(self, f):
+        """
+        Determine whether the passed file matches the Entity.
+
+        Args:
+            f (File): The BIDSFile instance to match against.
+
+        Returns: the matched value if a match was found, otherwise None.
+        """
+        if self.map_func is not None:
+            val = self.map_func(f)
+        else:
+            m = self.regex.search(f.path)
+            val = m.group(1) if m is not None else None
+
+        return self._astype(val)
+
+    def add_file(self, filename, value):
+        """ Adds the specified filename to tracking. """
+        self.files[filename] = value
+
+    def unique(self):
+        """ Returns all unique values/levels for the current entity. """
+        return list(set(self.files.values()))
+
+    def count(self, files=False):
+        """ Returns a count of unique values or files.
+
+        Args:
+            files (bool): When True, counts all files mapped to the Entity.
+                When False, counts all unique values.
+        Returns: an int.
+        """
+        return len(self.files) if files else len(self.unique())
+
+    def _astype(self, val):
+        if val is not None and self.dtype is not None:
+            val = self.dtype(val)
+        return val
+        
+
+class BIDSFile(object):
+    """ Represents a single file in a BIDS project.
+    
+    Args:
+        filename (str): Full path to file.
+        parent (BIDSNode): Optional parent node/directory.
+    
+    """
+    def __init__(self, filename, parent=None):
+        self.path = filename
+        self.filename = os.path.basename(self.path)
+        self.dirname = os.path.dirname(self.path)
+        self.tags = []
+        self.entities = {}
+        self.parent = parent
+
+    def _matches(self, entities=None, extensions=None, regex_search=False):
+        """
+        Checks whether the file matches all of the passed entities and
+        extensions.
+
+        Args:
+            entities (dict): A dictionary of entity names -> regex patterns.
+            extensions (str, list): One or more file extensions to allow.
+            regex_search (bool): Whether to require exact match (False) or
+                regex search (True) when comparing the query string to each
+                entity.
+        Returns:
+            True if _all_ entities and extensions match; False otherwise.
+        """
+        if extensions is not None:
+            if isinstance(extensions, six.string_types):
+                extensions = [extensions]
+            extensions = map(re.escape, extensions)
```

Comment:
What about dropping the above guard and using:

```suggestion
            extensions = map(re.escape, listify(extensions))
```